### PR TITLE
Fix degree bug, and add benchmarks of polynomial add/sub/neg

### PR DESCRIPTION
**Pull request by ValarDragon**
## Description

Parallelize `Add`, `Sub`, `AddAssign`, `SubAssign`, `Neg` operations on DensePolynomial.

Furthermore fix bugs in degree calculation when you have dense polynomial addition / subtraction that causes the leading coefficient to be 0.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by Pratyush** [comment]
Hmm I wonder if parallelizing these cheap operations would give better speedups if we used [this](https://docs.rs/rayon/1.5.0/rayon/iter/trait.IndexedParallelIterator.html#method.with_min_len) to increase the amount of computation on each thread

**Comment by ValarDragon** [comment]
Not sure I see a reason as to why that would speed this up, but I can add benchmarks for these operations to test it.

**Comment by ValarDragon** [comment]
Based on benchmarks for small sized polynomials, definitely appears like we should be using that for small polynomials. Results of addition speeds on the benchmark server:
```
"bls12_381" - add_polynomial/4
                        time:   [220.19 us 221.88 us 223.44 us]
"bls12_381" - add_polynomial/16
                        time:   [283.37 us 285.74 us 287.92 us]
"bls12_381" - add_polynomial/64
                        time:   [191.32 us 195.84 us 199.98 us]
"bls12_381" - add_polynomial/256
                        time:   [156.59 us 157.25 us 157.87 us]
```

Should we block merging this PR until we update `ark_std` for this?

**Comment by ValarDragon** [comment]
Hrmm, looks like `with_min_len` doesn't really fix the problem.

We should probably change add an ark_std macro for  `cfg_iter_with_min_size!(v, minimum_length)` being equivalent to 
```
#[cfg[feature(parallel)]]
if v.len() > minimum_length {
     cfg_iter!(v)
}
else {
    sequential code
}
```

**Comment by ValarDragon** [comment]
`with_min_len` benchmarks:

```
"bls12_381" - add_polynomial/1
                        time:   [95.391 us 96.408 us 97.278 us]
                        change: [-9.8377% -8.8342% -7.9433%] (p = 0.00 < 0.05)
"bls12_381" - add_polynomial/4
                        time:   [174.42 us 175.61 us 176.70 us]
                        change: [-11.489% -9.2890% -7.0652%] (p = 0.00 < 0.05)
bls12_381" - add_polynomial/16
                        time:   [223.92 us 225.63 us 227.20 us]
                        change: [-19.895% -18.767% -17.660%] (p = 0.00 < 0.05)
"bls12_381" - add_polynomial/64
                        time:   [212.25 us 216.25 us 219.94 us]
                        change: [+10.884% +13.254% +15.973%] (p = 0.00 < 0.05)
"bls12_381" - add_polynomial/256
                        time:   [244.27 us 247.81 us 251.16 us]
                        change: [+53.078% +55.495% +57.799%] (p = 0.00 < 0.05)
"bls12_381" - add_polynomial/1024
                        time:   [181.77 us 182.84 us 183.86 us]
                        change: [-15.485% -14.604% -13.712%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
"bls12_381" - add_polynomial/4096
                        time:   [199.15 us 202.84 us 206.36 us]
                        change: [-36.096% -34.938% -33.571%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
Benchmarking "bls12_381" - add_polynomial/16384: Collecting 100 samples                                                                         "bls12_381" - add_polynomial/16384
                        time:   [384.39 us 387.86 us 391.42 us]
                        change: [-33.035% -31.715% -30.393%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
Benchmarking "bls12_381" - add_polynomial/65536: Collecting 100 samples                                                                         "bls12_381" - add_polynomial/65536
                        time:   [965.96 us 975.14 us 985.47 us]
                        change: [-18.815% -14.592% -10.073%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe
Benchmarking "bls12_381" - add_polynomial/262144: Warming up for 3.0000                                                                         Benchmarking "bls12_381" - add_polynomial/262144: Collecting 100 samples                                                                        "bls12_381" - add_polynomial/262144
                        time:   [3.1964 ms 3.2494 ms 3.3045 ms]
```

**Comment by ValarDragon** [comment]
Hrmm, looks like `with_min_len` doesn't really fix the problem.

We should probably change add an ark_std macro for  `cfg_iter_with_min_size!(v, minimum_length)` being equivalent to 
```
#[cfg[feature(parallel)]]
if v.len() > minimum_length {
     cfg_iter!(v).with_min_len(minimum_length)
}
else {
    sequential code
}
```

And the minimum length being really large (like 2**20)?

Alternatively I can just undo the parallelization in this PR, and just copy over polynomial arithmetic infra, and bug fixes

**Comment by ValarDragon** [comment]
Something seems very wrong about these `add_polynomial` benchmarks. Polynomial evaluation which is N multiplications is 8x faster
```
"bls12_381" - add_polynomial/262144
                        time:   [3.1964 ms 3.2494 ms 3.3045 ms]
"bls12_381" - evaluate_polynomial/262144
                        time:   [406.14 us 408.14 us 410.51 us]
```

**Comment by Pratyush** [comment]
Hmm I think the overhead of parallelization is causing issues here.

I think for these ops it's fine to leave them as not parallel, because they'll be dominated by other stuff anyway.

**Comment by ValarDragon** [comment]
Yeah, that sounds good to me. I'll remove the parallelization, but keep the bug fixes + benchmarks

File: poly/src/polynomial/univariate/dense.rs [hunk 1]
```diff
@@ -242,15 +242,23 @@ impl<'a, 'b, F: Field> Add<&'a DensePolynomial<F>> for &'b DensePolynomial<F> {
...
-            for (a, b) in result.coeffs.iter_mut().zip(&self.coeffs) {
-                *a += b
-            }
+            result
+                .coeffs
+                .iter_mut()
+                .zip(&self.coeffs)
+                .for_each(|(a, b)| {
+                    *a += b;
+                });
```
Comment (ValarDragon) [inline poly/src/polynomial/univariate/dense.rs:261]: Should I undo changes like this?

Comment (Pratyush) [reply poly/src/polynomial/univariate/dense.rs:261]: nah this is fine

**Review by Pratyush — APPROVED** [review]

**Pull request merged**